torch/lib/THD/master_worker/master/generic/THDTensorMath.cpp

```diff
@@ -0,0 +1,311 @@
+  );
+}
+
+// void THDTensor_(randperm)(THDTensor *r_, THDGenerator *_generator, long n) {
+//   throw std::runtime_error("randperm is not available yet");
+//   THArgCheck(n > 0, 1, "must be strictly positive");
+
+//   THDTensor_(resize1d)(r_, n);
+
+// }
+
+void THDTensor_(reshape)(THDTensor *r_, THDTensor *t, THLongStorage *size) {
+  THDTensor_(resize)(r_, size, NULL);
+  //THDTensor_(copy)(r_, t); TODO
+}
+
+void THDTensor_(sort)(THDTensor *rt_, THDLongTensor *ri_,
+                             THDTensor *t, int dimension,
+                             int descendingOrder) {
+  THArgCheck(dimension >= 0 && dimension < THDTensor_(nDimension)(t),
+      2, "invalid dimension %d", dimension + TH_INDEX_BASE);
+
+  THDTensor_(resizeAs)(rt_, t);
+  // THDTensor_(copy)(rt_, t); TODO
+
+  {
+    THLongStorage *size = THDTensor_(newSizeOf)(t);
+    THDLongTensor_resize(ri_, size, NULL);
+    THLongStorage_free(size);
+  }
+
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorSort, rt_, ri_, t, dimension, descendingOrder),
+    THDState::s_current_worker
+  );
+}
+
+void THDTensor_(topk)(THDTensor *rt_, THDLongTensor *ri_,
+                      THDTensor *t, long k, int dim,
+                      int dir, int sorted) {
+  int numDims = THDTensor_(nDimension)(t);
+  THArgCheck(dim >= 0 && dim < numDims, 3, "dim not in range");
+
+  long sliceSize = THDTensor_(size)(t, dim);
+  THArgCheck(k > 0 && k <= sliceSize, 2, "k not in range for dimension");
+
+  THLongStorage *topKSize = THDTensor_(newSizeOf)(t);
+  THLongStorage_set(topKSize, dim, k);
+  THDTensor_(resize)(rt_, topKSize, NULL);
+  THDLongTensor_resize(ri_, topKSize, NULL);
+  THLongStorage_free(topKSize);
+
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorTopk, rt_, ri_, t, k, dim, dir, sorted),
+    THDState::s_current_worker
+  );
+}
+
+void THDTensor_(tril)(THDTensor *r_, THDTensor *t, long k) {
+  THArgCheck(THDTensor_(nDimension)(t) == 2, 1, "expected a matrix");
+
+  THDTensor_(resizeAs)(r_, t);
+
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorTril, r_, t, k),
+    THDState::s_current_worker
+  );
+}
+
+void THDTensor_(triu)(THDTensor *r_, THDTensor *t, long k) {
+  THArgCheck(THDTensor_(nDimension)(t) == 2, 1, "expected a matrix");
+
+  THDTensor_(resizeAs)(r_, t);
+
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorTriu, r_, t, k),
+    THDState::s_current_worker
+  );
+}
+void THDTensor_(cat)(THDTensor *r_, THDTensor *ta, THDTensor *tb, int dimension) {
+  THDTensor* inputs[2];
+  inputs[0] = ta;
+  inputs[1] = tb;
+  THDTensor_(catArray)(r_, inputs, 2, dimension);
+}
+
+void THDTensor_(catArray)(THDTensor *result, THDTensor **inputs,
+                          int numInputs, int dimension) {
+  // TODO I'm not quite sure about this - it looks like the
+  // TH implementation only uses the functions that were
+  // defined previously, so there's no need to send an explicit
+  // message
```

Comment:
It's prob better to implement it as computing the metadata for the result + sending a cat message to the worker. It's going to be much cheaper that way (only one op). This matters especially for CUDA, where there are a number of optimized GPU kernels for `cat` and they won't be used if this function will look like this